tests/unit/test_async_2comp.c

```diff
@@ -0,0 +1,127 @@
+ *
+ * To run with valgrind, use this command:
+ * <pre>mpiexec -n 4 valgrind -v --leak-check=full --suppressions=../../../tests/unit/valsupp_test.supp
+ * --error-exitcode=99 --track-origins=yes ./test_intercomm3</pre>
+ *
+ */
+#include <pio.h>
+#include <pio_tests.h>
+
+/* Number of processors that will do IO. */
+#define NUM_IO_PROCS 2
+
+/* Number of computational components to create. */
+#define COMPONENT_COUNT 2
+
+/* The number of tasks this test should run on. */
+#define TARGET_NTASKS 4
+
+/* The name of this test. */
+#define TEST_NAME "test_intercomm3"
+
+/** Run Tests for Init_Intercomm
+ *
+ */
+int
+main(int argc, char **argv)
+{
+    int verbose = 1;
+
+    /** Zero-based rank of processor. */
+    int my_rank;
+
+    /** Number of processors involved in current execution. */
+    int ntasks;
+
+    /** Different output flavors. */
+    int flavor[NUM_FLAVORS] = {PIO_IOTYPE_PNETCDF, PIO_IOTYPE_NETCDF,
+			       PIO_IOTYPE_NETCDF4C, PIO_IOTYPE_NETCDF4P};
+
+    /** The ID for the parallel I/O system. */
+    int iosysid[COMPONENT_COUNT];
+
+    /** The ncid of the netCDF file. */
+    int ncid;
+
+    /** The ID of the netCDF varable. */
+    int varid;
+
+    /** Return code. */
+    int ret;
+
+    /** Index for loops. */
+    int fmt, d, d1, i;
+
+    /* Initialize test. */
+    if ((ret = pio_test_init(argc, argv, &my_rank, &ntasks, TARGET_NTASKS)))
+	ERR(ERR_INIT);
+
+    /* How many processors will be used for our IO and 2 computation components. */
+    int num_procs[COMPONENT_COUNT + 1] = {2, 1, 1};
+
+    /* Is the current process a computation task? */
+    int comp_task = my_rank < 2 ? 0 : 1;
+
+    /* Index of computation task in iosysid array. Varies by rank and
+     * does not apply to IO component processes. */
+    int my_comp_idx = comp_task ? my_rank - 2 : -1;
+
+    /* Initialize the IO system. */
+    if ((ret = PIOc_Init_Async(MPI_COMM_WORLD, NUM_IO_PROCS, NULL, COMPONENT_COUNT,
+			       num_procs, NULL, iosysid)))
+	ERR(ERR_AWFUL);
+
+    /* All the netCDF calls are only executed on the computation
+     * tasks. The IO tasks have not returned from PIOc_Init_Intercomm,
+     * and when the do, they should go straight to finalize. */
+    if (comp_task)
+    {
+    	for (int flv = 1; flv < NUM_FLAVORS - 2; flv++)
+    	{
+	    char filename[NC_MAX_NAME + 1];
+
+	    /* Create a filename. */
+	    int sample = 1;
+	    sprintf(filename, "%s_%s_%d_%d.nc", TEST_NAME, flavor_name(flv), sample, my_comp_idx);
+
+	    /* Create sample file 1. */
+	    printf("%d %s creating file %s\n", my_rank, TEST_NAME, filename);
+	    if ((ret = create_nc_sample_1(iosysid[my_comp_idx], flavor[flv], filename, my_rank)))
+		ERR(ret);
+
+    	    /* Check the file for correctness. */
+    	    if ((ret = check_nc_sample_1(iosysid[my_comp_idx], flavor[flv], filename, my_rank)))
+    	    	ERR(ret);
+
+    	} /* next netcdf format flavor */
+
+	/* If I don't sleep here for a second, there are problems. */
+	sleep(2);
```

Comment:
This is a test that I am still working on. The sleep will certainly not remain in the code.

Right now this test is turned off. Once I get it working reliably I will turn it on again.